Document how to access metering data (#83)
appuio/appuio-io-docs: docs/modules/ROOT/pages/explanation/decisions/access-metering-data.adoc

= Access metering data to run reports

== Problem

> As outlined in xref:appuio-cloud:ROOT:references/architecture/metering.adoc[Metering of resource usage], metering data is collected into an instance of Thanos (Prometheus API).

Reviewer: ```suggestion
As discussed in xref:appuio-cloud:ROOT:references/architecture/metering.adoc[Metering of resource usage], metering data is collected into a https://thanos.io[Thanos] instance.
Thanos provides a query API which is guaranteed to be compatible with the Prometheus 2.x API.
```

> That data is supposed to be used to generate invoices.

Reviewer: ```suggestion
The collected data is intended to be used to generate invoices.
```

> It might also be used to show cost reports on the {portal}.

Reviewer: ```suggestion
In addition, the data may also be used to show cost reports on the {portal}.
```

> The Prometheus data is missing information, for example prices and discounts.

Reviewer: ```suggestion
It's important to note that the Prometheus data is missing information, for example prices and discounts.
```

> This data has to be augmented using another source.

Reviewer: ```suggestion
Therefore, the collected metering data has to be augmented with information provided by another source.
```

== Proposals

. Directly use the Prometheus API
+
The Prometheus API allows to get results for a single point in time.
It also allows to query all results within a given time range with a data point as defined by a given interval.

. Use of an intermediate storage
+
The Prometheus data is queried regularly, aggregated, enriched with product data and stored into an intermediate database.
That intermediate storage could then also contain data about pricing and discounts.

== Decision

Use of an intermediate storage

== Rationale

Prometheus queries are restricted in how many data points can be processed.
The queries required for the billing model are quite complex.

> It's not possible to run them once for a who billing interval (currently one month).

Reviewer: ```suggestion
It's not possible to run them once for a whole billing interval (currently one month).
```
